.github/workflows/docker.yaml

```diff
@@ -16,28 +9,40 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
+      - name: Extract branch name
+        if: github.event_name != 'pull_request'
+        run: |
+          REF=${GITHUB_REF#refs/*/}
+          if [ "$REF" == "master" ]; then
+            IMAGE_TAG=latest
+          else
+            # Replace '/' with '__'
+            IMAGE_TAG=${REF//\//__}
+          fi
+          echo "IMAGE_TAG=$IMAGE_TAG" >> $GITHUB_ENV
+      - name: Extract branch name
+        if: github.event_name == 'pull_request'
+        run: |
+          echo "IMAGE_TAG=$GITHUB_HEAD_REF" >> $GITHUB_ENV
+
+      # extract branch name on pull request
+      - name: Print branch name
+        run: echo "${IMAGE_TAG}"
+        
+      - name: Set env variables
+        run: |
+          echo "DOCKER_IMAGE=${DOCKER_IMAGE_BASENAME}:$IMAGE_TAG" >> $GITHUB_ENV
+
       - name: Login to DockerHub
-        if: github.event_name == 'push'
         uses: docker/login-action@v1
         with:
           username: ${{ secrets.DOCKERHUB_USERNAME }}
           password: ${{ secrets.DOCKERHUB_TOKEN }}
 
-      - name: Get image name
-        run: |
-          VERSION=${GITHUB_REF#refs/*/}
-          if [ "$GITHUB_EVENT_NAME" == "pull_request" ]; then
-            IMAGE_TAG=$GITHUB_SHA
-          elif [ "$VERSION" == "master" ]; then
-            IMAGE_TAG=latest
-          elif [ "$VERSION" == "es7" ]; then
-            IMAGE_TAG=latest-es7
-          else
-            IMAGE_TAG=$VERSION
-          fi
-          echo "DOCKER_IMAGE=$DOCKER_IMAGE_BASENAME:$IMAGE_TAG" >> $GITHUB_ENV
-
       - run: docker build --label "org.label-schema.vcs-ref=$GITHUB_SHA" -t $DOCKER_IMAGE .
-
-      - if:  github.event_name == 'push'
-        run: docker push $DOCKER_IMAGE
+      
+      - name: Docker push
+        if: github.event_name == 'push' || github.event_name == 'pull_request'
+        run: |
+          docker push $DOCKER_IMAGE
```

Comment:
Is that a good practice to push a docker image for all branches by default? I feel like we will be polluting Docker Hub a lot.

In the case of Idunn this is not a big issue as the last layers mostly contain Python sources, so we only upload a few megabytes. But on other repositories the cost can be couples of dozens of megabytes if there are compiled binaries or big raw data in the image. 

The only reason why this could look ok to me in a more general situation is that we are using a free Docker Hub account for which [inactive images appears to be removed after 6 months](https://github.com/docker/roadmap/issues/152).

Reply:
is a good practice to test a feature before merge to master